api/server/volume.go

```diff
@@ -362,62 +330,85 @@ func (vd *volAPI) volumeSet(w http.ResponseWriter, r *http.Request) {
 	if req.Options["UNMOUNT_BEFORE_DETACH"] == "true" {
 		detachOptions.UnmountBeforeDetach = true
 	}
-	mountAttachClient := api.NewOpenStorageMountAttachClient(conn)
 
-	for err == nil && req.Action != nil {
-		if req.Action.Attach != api.VolumeActionParam_VOLUME_ACTION_PARAM_NONE {
-			if req.Action.Attach == api.VolumeActionParam_VOLUME_ACTION_PARAM_ON {
-				_, err = mountAttachClient.Attach(ctx, &api.SdkVolumeAttachRequest{
-					VolumeId:      volumeID,
-					Options:       attachOptions,
-					DriverOptions: req.GetOptions(),
-				})
-			} else {
-				_, err = mountAttachClient.Detach(ctx, &api.SdkVolumeDetachRequest{
-					VolumeId:      volumeID,
-					Options:       detachOptions,
-					DriverOptions: req.GetOptions(),
-				})
+	unmountOptions := &api.SdkVolumeUnmountOptions{}
+	if req.Options["DELETE_AFTER_UNMOUNT"] == "true" {
+		unmountOptions.DeleteMountPath = true
+	}
+	if req.Options["WAIT_BEFORE_DELETE"] == "true" {
+		unmountOptions.NoDelayBeforeDeletingMountPath = false
+	} else {
+		unmountOptions.NoDelayBeforeDeletingMountPath = true
+	}
+
+	if req.Locator != nil || req.Spec != nil {
+		// Only update spec if spec and locator are not nil.
+		vol, err := volumes.Inspect(ctx, &api.SdkVolumeInspectRequest{
```

Comment:
You have to be careful on the use of `err`. Here, this `err` is a new one, but on line 379 and on, it is the one on the previous scope.

Then `err` from the top scope it is evaluated on 423.

Reply:
Yeah that's okay. This scope is just for updating the volumes.

If the Inspect fails and the err is other than not found, it returns with error anyways.

Line 423 should be using `err` from the top scope as that reflects the actual attach, mount, detach and unmount operations.